### PR TITLE
Preset fixes for unit restrictions

### PR DESCRIPTION
**Pull request by HUSSAR-mtrela**
This PR adds:
- fixes made by @hussar-mtrela for issues: #2512, #2083, #2515
- fixes made by @shalkya for issues: #2546 #2547 #2548 #2549 #2550 #2551 #2552 #2553 #2554 #2555 #2556 #2557 #2558 #2559 #2560 #2561 #2562 #2563 #2564 #2565, #2518, 
- and a few other minor fixes in presets of unit restrictions

Note that this PR includes all fixes made in PR #2452 and meets requested changes in discussion of that PR. 

This PR is ready for testing.

### CODE REVIEW

**Comment by shalkya** [comment]
```-    STEALTH_AIR  = "((STEALTH + COUNTERINTELLIGENCE) * AIR)",
-    STEALTH_LAND = "((STEALTH + COUNTERINTELLIGENCE) * LAND + url0301_Stealth + url0301_Cloak)",
-    STEALTH_NAVY = "((STEALTH + COUNTERINTELLIGENCE) * NAVAL)",
+    STEALTH_AIR  = "((STEALTH + ura0302 + ura0303 + ura0304) * AIR)",
+    STEALTH_LAND = "((STEALTH + url0402 + url0306) * LAND + url0301_Stealth + url0301_Cloak)",
+    STEALTH_NAVY = "((STEALTH + COUNTERINTELLIGENCE) * NAVAL) + xrs0205",```
shouldn't you add counter-intelligent tag to cyb strat/asf, mermaid and remove it from rhino, beetle, loya, brick, selen to keep it less hardcoded (more dynamic) and in a less hacky way ?

**Comment by shalkya** [comment]
this change the order of cybran torpedo defense sorting : 
image
used to be T1 then T2 then T3
now it's T3 then T1 then T2

image
it also allow hives, while still disabling kennels. Should disable hives too.

image
it allows SCU when T3 are disable. The SCU should be disable. 

image
paragon and novax (the building) are allowed. While they should be disable.

**Comment by HUSSAR-mtrela** [comment]
I will fix issues listed above in this PR but sorting of units will be fixed in a new PR because I have other similar issue to fix and I dont want to bundle to many fixes in one PR. 
I think the `COUNTERINTELLIGENCE` category is used for some other purpose than indicated that a unit has a stealth field. So adding it to all stealth units is not a good idea. We really need new categories: `STEALTHFIELDPACK` and `STEALTHPERSONALPACK` or better yet the Blueprint.lua should insert missing categories when blueprints are being loaded. This way, FAF units as well as modded units will get correct categories.

**Comment by HUSSAR-mtrela** [comment]
the No T3 restriction work as expected because SCUs are engineers and the restriction excludes all engineers as it is explained in the tooltip.

**Comment by HUSSAR-mtrela** [comment]
@shalkya  which restriction are you referring to by asking this?
> would it be possible to set up a warning in the description of no anti-air, so that people know that drone are still buildable

**Comment by shalkya** [comment]
well yeah i consider scu mostly as combat unit than engineer, hence why i putted that here.

**Comment by HUSSAR-mtrela** [comment]
Hi @shalkya I fixed all issues that you listed above except of the sorting of cybran torpedo defense structures because I needs more work and I plan to fix it in a new PR.

File: lua/ui/lobby/UnitsRestrictions.lua [hunk 1]
```diff
@@ -47,7 +47,7 @@ Expressions = {
     -- excluding engineers, economy, and factories because players will not progress to higher tech levels
     T1          = "(TECH1 - (MOBILE * ENGINEER) - (STRUCTURE * FACTORY) - ECONOMIC)",
     T2          = "(TECH2 - (MOBILE * ENGINEER) - (STRUCTURE * FACTORY) - ECONOMIC)",
-    T3          = "(TECH3 - (MOBILE * ENGINEER) - (STRUCTURE * FACTORY) - ECONOMIC)",
+    T3          = "(TECH3 - (MOBILE * ENGINEER) - (STRUCTURE * FACTORY) - ECONOMIC + SUBCOMMANDER)",
```
Comment (shalkya) [inline lua/ui/lobby/UnitsRestrictions.lua]: it didn't remove the scu in my testing

**Comment by FlyingThunder** [comment]
Addition: On the "No air" preset, some form of anti air HAS TO BE ENABLED! (either MAA or T1 stationary), because if not, players can abuse the UEF drones to e.g. reclaim enemy bases without them being able to counter it!

**Comment by shalkya** [comment]
we already had this discussion, there is a "no drones" already possible in another preset. If you want no air and no drones, you should go by selecting the 2 presets. 
the AA are already enable in this pr when "no air" is on.

**Comment by shalkya** [comment]
i've been adding a commit that does change plenty of restriction, tell me if you prefer that i split each change in different commit, or if you can review it without too much issue like that.
Also there is dubious change like snoop removed from direct fire unit. You can tell me what you think about it.
Finally i introduced some bug similar to the one of torpedo launcher (that hussar introduced), ie : unit changed their spot in the restriction manager.
I will let hussar handle this issue.

File: lua/ui/lobby/UnitsRestrictions.lua [hunk 2]
```diff
@@ -70,17 +70,16 @@ Expressions = {
...
     TMLMOBILE   = "(MOBILE * LAND * INDIRECTFIRE * SILO)", -- XSL0111 + URL0111 + UEL0111 + UAL0111 + XEL0306
 
     -- added exclusion of engineers and structures because they are restricted by other presets
-    LAND        = "(LAND - ENGINEER - STRUCTURE)",
+    LAND        = "(LAND - ENGINEER - STRUCTURE + SUBCOMMANDER)",
     -- added restriction of air staging structures because they are not needed when all air units are restricted
     AIR         = "((STRUCTURE * AIRSTAGINGPLATFORM) + (AIR - POD))",
-    -- added restriction of anti-navy structures because they are not needed when all navy units are restricted
-    NAVAL       = "((STRUCTURE * ANTINAVY) + NAVAL - (MOBILESONAR * TECH3))",
+    NAVAL       = "NAVAL - STRUCTURE - (MOBILESONAR * TECH3))",
```
Comment (HUSSAR-mtrela) [inline lua/ui/lobby/UnitsRestrictions.lua]: @Strogoo why did you remove anti-navy structures from No Naval restriction?

File: lua/ui/lobby/UnitsRestrictions.lua [hunk 3]
```diff
@@ -117,33 +116,33 @@ Expressions = {
     T2_BASE_SPAM  = "(TECH2 * STRUCTURE - FACTORY - MASSEXTRACTION - MASSSTORAGE - MASSFABRICATION - ENERGYPRODUCTION - ENERGYSTORAGE)",
     T3_BASE_SPAM  = "(TECH3 * STRUCTURE - FACTORY - MASSEXTRACTION - MASSSTORAGE - MASSFABRICATION - ENERGYPRODUCTION - ENERGYSTORAGE + MOBILESONAR)",
 
-    -- including Soul Ripper, Czar, and Ahwassa
-    SNIPES_AIR    = "(AIR * (TECH2 + TECH3) * (GROUNDATTACK + BOMBER + ANTINAVY) + ura0401 + uaa0310 + xsa0402)",
+    -- including Satellite, Soul Ripper, Czar, and Ahwassa and GhetoGunship
+    SNIPES_AIR    = "(AIR * (TECH2 + TECH3 + EXPERIMENTAL) * (GROUNDATTACK + BOMBER + ANTINAVY) + TRANSPORTATION) + uaa0310", -- CZAR
```
Comment (HUSSAR-mtrela) [inline lua/ui/lobby/UnitsRestrictions.lua]: @shalkya please restrict labs/mech units instead of `TRANSPORTATION` units because transport units are essential to quickly expand

File: lua/ui/lobby/UnitsRestrictions.lua (hunk 2 above)
Comment (shalkya) [reply lua/ui/lobby/UnitsRestrictions.lua]: because there could be a use for anti-navy structure, ie : defending against amphibious unit.

File: lua/ui/lobby/UnitsRestrictions.lua (hunk 3 above)
Comment (shalkya) [reply lua/ui/lobby/UnitsRestrictions.lua]: what would be great, would be to remove the ability to fire from transport of lab. But i guess that if it's not possible/too complicated, removing labs would be the best idea

Comment (HUSSAR-mtrela) [reply lua/ui/lobby/UnitsRestrictions.lua]: please use labs/mech unit instead of Transport units and your commit should be good

**Comment by shalkya** [comment]
there is still some issue around : can't remove quantum gateway for "no T3" not needed since we don't allow scu to be build
can't allow factory to be build under "no air" restriction

there is some wrong description of restriction, out of my head : "no T1 air" description i can't update for some reason; also "no T3"

and also we introduced the same bug of harms changing spot in unit manager to other unit like some T1 maa, selen, etc...

**Comment by shalkya** [comment]
fix :
#2546 #2547 #2548 #2549 #2550 #2551 #2552 #2553 #2554 #2555 #2556 #2557 #2558 #2559 #2560 #2561 #2562 #2563 #2564 #2565

**Comment by HUSSAR-mtrela** [comment]
we already have issue reported for harms changing spot in unit manager and I will fix it. We should not bundle not related fixes in the same PR (e.g. UI fixes and restriction fixes)

Also, I need to keep quantum gateway for "no T3" restriction because many mods add new/air units that are build from gateway. I disagree with allowing factories to be build under "no air" restriction or under "no navy" restriction because they are useless in those scenarios and Land factory is the best for producing engineers.

**Comment by HUSSAR-mtrela** [comment]
@Strogoo I made a few additional corrections and this PR should be ready for merging now. Please see my previous comment/answer to your questions.

**Comment by HUSSAR-mtrela** [comment]
Thanks @shalkya
